waveorder/filter.py

```diff
@@ -0,0 +1,202 @@
+    have extent 1/dx. Note that there is no need for io_filter_bank to have
+    sample spacing 1/(n*dx) because io_filter_bank will be stretched.
+
+    Parameters
+    ----------
+    io_filter_bank : torch.Tensor
+        The filter bank to be applied in the frequency domain.
+        The spatial extent of io_filter_bank must be 1/dx, where dx is the
+        sample spacing of i_input_array.
+
+        Leading dimensions are the input and output dimensions.
+        io_filter_bank.shape[:2] == (num_input_channels, num_output_channels)
+
+        Trailing dimensions are spatial frequency dimensions.
+        io_filter_bank.shape[2:] == (Z', Y', X') or (Y', X')
+
+    i_input_array : torch.Tensor
+        The real-valued input array with sample spacing dx to be filtered.
+
+        Leading dimension is the input dimension, matching the filter bank.
+        i_input_array.shape[0] == i
+
+        Trailing dimensions are spatial dimensions.
+        i_input_array.shape[1:] == (Z, Y, X) or (Y, X)
+
+    Returns
+    -------
+    torch.Tensor
+        The filtered real-valued output array with shape
+        (num_output_channels, Z, Y, X) or (num_output_channels, Y, X).
+
+    """
+
+    # Ensure all dimensions of transfer_function are smaller than or equal to input_array
+    if any(
+        t > i
+        for t, i in zip(io_filter_bank.shape[2:], i_input_array.shape[1:])
+    ):
+        raise ValueError(
+            "All spatial dimensions of io_filter_bank must be <= i_input_array."
+        )
+
+    # Ensure the number of spatial dimensions match
+    if io_filter_bank.ndim - i_input_array.ndim != 1:
+        raise ValueError(
+            "io_filter_bank and i_input_array must have the same number of spatial dimensions."
+        )
+
+    # Ensure the input dimensions match
+    if io_filter_bank.shape[0] != i_input_array.shape[0]:
+        raise ValueError(
+            "io_filter_bank.shape[0] and i_input_array.shape[0] must be the same."
+        )
+
+    num_input_channels, num_output_channels = io_filter_bank.shape[:2]
+    spatial_dims = io_filter_bank.shape[2:]
+
+    # Pad input_array until each dimension is divisible by transfer_function
+    pad_sizes = [
+        (0, (t - (i % t)) % t)
+        for t, i in zip(
+            io_filter_bank.shape[2:][::-1], i_input_array.shape[1:][::-1]
+        )
+    ]
+    flat_pad_sizes = list(itertools.chain(*pad_sizes))
+    padded_input_array = torch.nn.functional.pad(i_input_array, flat_pad_sizes)
+
+    # Apply the transfer function in the frequency domain
+    fft_dims = [d for d in range(1, i_input_array.ndim)]
+    padded_input_spectrum = torch.fft.fftn(padded_input_array, dim=fft_dims)
+
+    # Matrix-vector multiplication over f
+    # If this is a bottleneck, consider extending `stretched_multiply` to
+    # a `stretched_matrix_multiply` that uses an call like
+    # torch.einsum('io..., i... -> o...', io_filter_bank, padded_input_spectrum)
+    padded_output_spectrum = torch.zeros(
```

Comment:
Does this need a `device=padded_input_spectrum.device` to work on the GPU?

Reply:
Good point. Haven't tested end-to-end yet, but can't hurt. Fixed.